private/bufpkg/bufplugin/bufplugindocker/docker.go

```diff
@@ -0,0 +1,203 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufplugindocker
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"net"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufplugin/bufpluginconfig"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stringid"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type Client interface {
+	// Build creates a Docker image for the plugin using the Dockerfile.plugin and plugin config.
+	Build(ctx context.Context, dockerfile io.Reader, config *bufpluginconfig.Config, options ...BuildOption) (*BuildResponse, error)
+	// Push the Docker image to the remote registry.
+	Push(ctx context.Context, image string, auth RegistryAuth) (*PushResponse, error)
+	// Delete removes the Docker image from local Docker Engine.
+	Delete(ctx context.Context, image string) (*DeleteResponse, error)
+	// Close releases any resources used by the underlying Docker client.
+	Close() error
+}
+
+type BuildOption func(*buildParams)
+
+func WithConfigDirPath(path string) BuildOption {
+	return func(params *buildParams) {
+		params.ConfigDirPath = path
+	}
+}
+
+type BuildResponse struct {
+	// Image contains the Docker image name in the local Docker engine including the tag (i.e. plugins.buf.build/library/some-plugin:<id>, where <id> is a random id).
+	// It is created from the bufpluginconfig.Config's Name.IdentityString() and a unique id.
+	Image string
+	// Digest specifies the Docker image digest in the format <hash_algorithm>:<hash>.
+	// Example: sha256:65001659f150f085e0b37b697a465a95cbfd885d9315b61960883b9ac588744e
+	Digest string
+}
+
+type PushResponse struct{}
+
+type DeleteResponse struct{}
+
+type dockerAPIClient struct {
+	cli    *client.Client
+	logger *zap.Logger
+}
+
+var _ Client = (*dockerAPIClient)(nil)
+
+func (d *dockerAPIClient) Build(ctx context.Context, dockerfile io.Reader, pluginConfig *bufpluginconfig.Config, options ...BuildOption) (*BuildResponse, error) {
+	params := &buildParams{}
+	for _, option := range options {
+		option(params)
+	}
+	// TODO: Need to determine how contextDir parameter is used in Docker engine.
+	buildkitSession, err := createSession(ctx, fmt.Sprintf("%s/%s", pluginConfig.Name.Owner(), pluginConfig.Name.Plugin()), params.ConfigDirPath, zap.L())
+	if err != nil {
+		return nil, err
+	}
+
+	dockerContext, err := createDockerContext(dockerfile)
+	if err != nil {
+		return nil, err
+	}
+
+	eg, errGroupCtx := errgroup.WithContext(ctx)
```

Comment:
If we really _must_ use `errgroup` here, we should explain why. Otherwise, we should use `pkg/thread`. I know you had a good reason for this, but let's make sure it's loud and clear.